[RISCV] Lower SELECT's with one constant more efficiently using Zicond

bababuck opened the pull request:
See #143580 for MR with the test commit.

Performs the following transformations:
(select c, c1, t) -> (add (czero_nez t - c1, c), c1)
(select c, t, c1) -> (add (czero_eqz t - c1, c), c1)


@mgudim

bababuck commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -9073,19 +9078,36 @@ SDValue RISCVTargetLowering::lowerSELECT(SDValue Op, SelectionDAG &DAG) const {
+        IsCZERO_NEZ = isa<ConstantSDNode>(TrueV);
+        ConstVal = IsCZERO_NEZ ? TrueV : FalseV;
+
+        SDValue &RegV = IsCZERO_NEZ ? FalseV : TrueV;
```
Why not? Won't it be the same functionally but possibly faster with a reference here?

mshockwave replied:
SDValue should generally be passed by value because it's a simple value that uses a pointer and an unsigned underlying, which, at least in modern architectures, will be passed by two registers. On the other hand pass by reference will guarantee to generate a memory access.

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -9073,19 +9078,36 @@ SDValue RISCVTargetLowering::lowerSELECT(SDValue Op, SelectionDAG &DAG) const {
+      // The one constant case is only efficient is the constant fits into
+      // `ADDI`
+      if (BothAreConstant ||
+          isInt<12>(dyn_cast<ConstantSDNode>(ConstVal)->getSExtValue())) {
```
We should have a test for this case

bababuck replied:
I added a test for -2048, would you want one for `2048` as well?

bababuck replied:
Ended up adding a test for 2048 as well. How do you find the balance of not enough tests and too many tests (where it becomes impossible to manually verify the changes are expected).

bababuck commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -9088,6 +9088,22 @@ SDValue RISCVTargetLowering::lowerSELECT(SDValue Op, SelectionDAG &DAG) const {
+      bool IsCZERO_NEZ = isa<ConstantSDNode>(TrueV);
+      SDValue ConstVal = IsCZERO_NEZ ? TrueV : FalseV;
+      // Efficient only if the constant and its negation fit into `ADDI`
+      if (std::abs(cast<ConstantSDNode>(ConstVal)->getSExtValue()) < 0x800) {
```
Can we just use XOR for all the cases (assuming `isInt<12>`)?

topperc replied:
Mathematically, yes, but ADDI is more compressible.

bababuck commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -9088,6 +9088,31 @@ SDValue RISCVTargetLowering::lowerSELECT(SDValue Op, SelectionDAG &DAG) const {
+                        DL, VT, SubOp, CondV);
+        return DAG.getNode(ISD::ADD, DL, VT, CMOV, ConstVal);
+      }
+      // Can use XORI if Const == -0x800
+      else if (RawConstVal == -0x800) {
+        SDValue XorOp = DAG.getNode(ISD::XOR, DL, VT, RegV, ConstVal);
+        SDValue CMOV =
+            DAG.getNode(IsCZERO_NEZ ? RISCVISD::CZERO_NEZ : RISCVISD::CZERO_EQZ,
+                        DL, VT, XorOp, CondV);
+        return DAG.getNode(ISD::XOR, DL, VT, CMOV, ConstVal);
+      }
```
@topperc Is this too verbose to have this seperate clause? I could pull out the meat of these two clauses into a function that I pass the opcodes too, but that felt like it would be hard to read.

topperc reviewed (approved):
LGTM

mgudim commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -9088,6 +9088,32 @@ SDValue RISCVTargetLowering::lowerSELECT(SDValue Op, SelectionDAG &DAG) const {
+      }
+      // Efficient only if the constant and its negation fit into `ADDI`
+      // Prefer Add/Sub over Xor since can be compressed for small immediates
+      if (isInt<12>(RawConstVal)) {
```
do we need to check isInt<12>(-RawConstVal)?

bababuck replied:
I believe the only case that `isInt<12>(Val)`  is true and `isInt<12>(-Val)` is not is for `-2048`, which we handle before this (see https://github.com/llvm/llvm-project/pull/143581#discussion_r2138932485).

bababuck committed:
[RISCV] Efficiently lower select %cond, %constant, %variable using zicond extension

Provides the following lowerings, which reduces instruction count by 1 for rv64:
(select cond, const, t) -> (add (czero_nez t - c1, cond), const)
(select cond, t, const) -> (add (czero_eqz t - c1, cond), const)

For the special case of c1 == -2048
(select cond, -2048, t) -> (xor (czero_nez (xor t, -2048), cond), -2048)
(select cond, t, -2048) -> (xor (czero_eqz (xor t, -2048), cond), -2048)

bababuck commented:
Rebased to latest `main`, removed the test commit (since that was merged already).

bababuck commented:
@mshockwave am I good to merge?

mshockwave reviewed (approved):
LGTM
